### PR TITLE
feat(schema-registry): Check if kafka topic exists

### PR DESCRIPTION
**Pull request by frondeus**

### CODE REVIEW

File: schema-registry/src/error.rs [hunk 1]
```diff
@@ -50,6 +54,32 @@ impl From<jsonschema::CompilationError> for RegistryError {
     }
 }
 
+impl From<utils::messaging_system::CommunicationError> for RegistryError {
+    fn from(error: utils::messaging_system::CommunicationError) -> Self {
+        Self::MQError(error.to_string())
+    }
+}
+
+impl From<RegistryError> for Status {
```
Comment (Kononnable) [inline schema-registry/src/error.rs:63]: Maybe it's the way to go, but it looks ugly :P

File: schema-registry/src/rpc.rs [hunk 2]
```diff
@@ -90,6 +94,15 @@ impl SchemaRegistry for SchemaRegistryImpl {
             kafka_topic: request.topic_name,
         };
 
+        if !self
+            .mq_metadata
+            .topic_exists(&new_schema.kafka_topic)
```
Comment (USSAMOH2) [inline schema-registry/src/rpc.rs:99]: This should also be done when we update the schema topic.

File: utils/src/messaging_system/metadata_fetcher.rs [hunk 3]
```diff
@@ -0,0 +1,42 @@
+use std::time::Duration;
+
+use anyhow::Context;
+use rdkafka::{producer::BaseProducer, ClientConfig};
+
+use super::CommunicationResult;
+
+pub enum MetadataFetcher {
```
Comment (USSAMOH2) [inline utils/src/messaging_system/metadata_fetcher.rs:8]: I'm not sure if this needs to be generic, but you don't need to remove it if you don't want to.

**Review by USSAMOH2 — DISMISSED** [review]

File: utils/src/messaging_system/metadata_fetcher.rs [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    pub async fn new_kafka(brokers: &str) -> CommunicationResult<Self> {
+        let producer = ClientConfig::new()
+            .set("bootstrap.servers", &brokers)
+            .create()
+            .context("Metadata fetcher creation failed")?;
+
+        Ok(Self::Kafka { producer })
+    }
+
+    pub async fn topic_exists(&self, topic: &str) -> CommunicationResult<bool> {
```
Comment (Kononnable) [inline utils/src/messaging_system/metadata_fetcher.rs:22]: If this is abstraction over messaging protocol we probably should use name like 'destination channel/pipe' or something similar - topic is related only to kafka.

File: utils/src/messaging_system/mod.rs [hunk 5]
```diff
@@ -4,31 +4,41 @@ pub mod consumer;
...
 
+pub mod metadata_fetcher;
+
 #[derive(Debug, Error)]
 pub enum CommunicationError {
     #[error("Error during communication via message queue \"{0}\"")]
     InternalError(String),
+
+    #[error("Error during joining blocking task \"{0}\"")]
+    TokioError(String),
```
Comment (Kononnable) [inline utils/src/messaging_system/mod.rs:15]: I would still say it's internal error - not related to application business logic.

File: utils/src/messaging_system/metadata_fetcher.rs (hunk 4 above)
Comment (frondeus) [reply utils/src/messaging_system/metadata_fetcher.rs:22]: But then why we are keeping `topic` inside of vertex :)?

File: utils/src/messaging_system/mod.rs (hunk 5 above)
Comment (frondeus) [reply utils/src/messaging_system/mod.rs:15]: Then i'd need to make an error message more generic cause its no longer `communication via message queue`

Comment (luke-biel) [reply utils/src/messaging_system/mod.rs:15]: Hmm, i think this means more to rename Internal somehow as opposed to merging these two

Comment (frondeus) [reply utils/src/messaging_system/mod.rs:15]: Now I'm thinking it's no longer `CommunicationError`

Comment (frondeus) [reply utils/src/messaging_system/mod.rs]: Changed it to... `Error` :)

**Review by Kononnable — APPROVED** [review]

**Pull request merged**